splitting WEBAPP_BUNDLES / Generate PWA

donsalari opened the pull request:
splitting WEBAPP_BUNDLES into PREFETCH_BUNDLES, WEBAPP_HEADER_BUNDLES and WEBAPP_BODY_BUNDLES

sample image is required bc of `https://github.com/electrode-io/electrode-archetype-react-app/blob/master/config/webpack/partial/pwa.js#L74`


## App generated with PWA
<img width="1476" alt="o1" src="https://cloud.githubusercontent.com/assets/4782871/20913516/70b6e928-bb2d-11e6-8651-d608666f899c.png">

jchip commented on generators/webapp/templates/server/plugins/webapp/index.js:
```diff
@@ -63,26 +89,36 @@ function makeRouteHandler(options, userContent) {
-      const jsLink = js ? `<script src="${js}"></script>` : "";
-      return `${cssLink}${jsLink}`;
+      const font = "https://fonts.googleapis.com/icon?family=Material+Icons|Open+Sans";
+      const fontLink = `<link rel="stylesheet" href="${font}" />`;
```
If this is only needed for PWA, then should add it only if there's a manifest.

donsalari replied:
yep makes sense. ill add the conditions

jchip commented:
another thing, what if user doesn't want to generate app with PWA?  What determines if manifest.json gets created or not?

donsalari commented:
@jchip I am adding a question on whether the user needs a pwa app. So separating out the functionality soon.

donsalari commented:
manifest.json will be created only if the user adds a `sw-config.json`, a service worker config.

donsalari commented:
@jchip there is no code specific to pwa. i have renamed the title and updated the description.
I'll add the pwa files as part of a separate question.

jchip commented on generators/webapp/templates/server/plugins/webapp/index.js:
```diff
@@ -21,33 +21,59 @@ function loadAssetsFromStats(statsFilePath) {
   /* Create a route handler */
+  /*eslint max-statements: 0*/
   return (request, reply) => {
+    if (global.navigator) {
```
please add a comment explaining what this is for.

donsalari replied:
its required by material-ui if we render server side so it has a better understanding of the screen size. its an optional param. ill take it off since material-ui is not needed for every project.

jchip commented on generators/webapp/templates/server/plugins/webapp/index.js:
```diff
@@ -63,26 +89,36 @@ function makeRouteHandler(options, userContent) {
     const callUserContent = (content) => {
       const x = content(request);
       return !x.catch ? x : x.catch((err) => {
-        return Promise.reject({
```
we want to render the page with the error message?  Then we might want to make it consistent to do the same in the catch below.  You can add this as an issue for later instead of this PR though.

jchip commented on generators/webapp/templates/server/plugins/webapp/index.js:
```diff
@@ -127,7 +167,7 @@ function makeRouteHandler(options, userContent) {
       })
       .catch((err) => {
-        reply(err.html).code(err.status || HTTP_ERROR_500);
+        reply(err.message).code(err.status || HTTP_ERROR_500);
```
removing the `Promise.reject` above renders the error there but here it is not.

jchip commented on generators/app/index.js:
```diff
@@ -231,12 +237,28 @@ module.exports = generators.Base.extend({
         this.destinationPath(f)
       );
     });
+
```
You can avoid copying the images and sw-register.js in the copyTpl using globOptions

```js
    this.fs.copyTpl(
      this.templatePath('client'),
      this.destinationPath('client'),
      { pwa: this.props.pwa },
      {}, // template options
      { // copy options
        globOptions: {
          ignore: [ "**/client/images/**", !this.props.pwa && "**/client/sw-register.js" ]
        }
      }
    );
```

donsalari replied:
sure. although `!this.props.pwa && '**/client/sw-register.js'` does not work. 
`TypeError: pattern.slice is not a function`
so i am ignoring both the files for now and then include them if its a pwa

jchip replied:
odd, it works for me. it may be require an actual string.  `!this.props.pwa && '**/client/sw-register.js' || ''`

donsalari replied:
yep that was it. worked like charm :)

jchip commented on generators/app/index.js:
```diff
@@ -148,6 +148,12 @@ module.exports = generators.Base.extend({
+        {
+          type: "confirm",
           name: "createDirectory",
           message: "Would you like to create a new directory for your project?",
```
this prompt should not appear again when you run `yo electrode` in an existing app.  Use the `when` property.

https://github.com/electrode-io/generator-electrode/blob/master/generators/app/index.js#L102

jchip commented on generators/app/index.js:
```diff
@@ -148,6 +148,12 @@ module.exports = generators.Base.extend({
         {
           type: "confirm",
+          name: "pwa",
+          message: "Would you like to make a Progressive Web App?",
```
do you want to ask this question again if user run `yo electrode` in an existing app?

jchip commented:
👍
